### ISSUE TITLE
Provide a mechanism to recover focus in menus with no valid focus search outcomes

### ISSUE THREAD

**Issue by pneuma163** [focustense/StardewUI#63]
This bug concerns how focusable and horizontal-content-alignment or vertical-content-alignment work together. The following sml tries to make the inner frame be focusable but the cursor is placed where it would be if the alignments were not present (or set to start) rather than at the center of the entire view where the inner frame is located as expected.
```
<frame layout="1280px 720px"
        background={@Mods/StardewUI/Sprites/ControlBorder}
        horizontal-content-alignment="middle"
        vertical-content-alignment="middle">
    <frame layout="320px 320px" 
            background={@Mods/StardewUI/Sprites/ControlBorder}
            focusable="true" />
</frame>
```

**Comment by focustense** [comment]
Afraid I don't understand what the issue is here. Focus behavior and alignment have no relationship to each other; alignment is part of layout, and focus happens post-layout, so there is no way for one to have any impact on the other. Also, focus has to do with movement between elements, and only one element is focusable here.

> the cursor is placed where it would be if the alignments were not present

Can you explain this in more detail? With what trigger, under what circumstances, etc.? Since the issue is not obvious to me, I think I will need this one of the form of precise repro steps and expected/actual outcomes. Or a [SSCCE](http://sscce.org).

**Comment by pneuma163** [comment]
For sure. That was a bit of a rush job anyway.

1. Create a menu:
```
Game1.activeClickableMenu = viewEngine.CreateMenuFromAsset(
    $"Mods/{MODID}/Views/Test",
    context);
```
from the following asset named Test.sml (different from the first post)
```
<frame layout="320px 640px">
    <lane orientation="vertical">
        <frame layout="320px 320px" 
                background={@Mods/StardewUI/Sprites/ControlBorder}
                focusable="true" />
        <frame layout="320px 320px" 
                background={@Mods/StardewUI/Sprites/ControlBorder}
                horizontal-content-alignment="middle"
                vertical-content-alignment="middle" >
            <button layout="100px 100px" />
        </frame>
    </lane>
</frame>
```

2. Set Gamepad Mode in the in game options to Force On and display the above menu in game (I do not have a controller so I'm only detailing my report in the way I can recreate it). This creates a menu of two vertically stacked frames with a button in the center of the bottom one.
3. Move the cursor using WASD on a keyboard and notice where the cursor is placed within the two frames in the vertical lane.

I am including an image for reference
![menu](https://github.com/user-attachments/assets/d21fd179-d76d-4abc-8f9f-55945b495f40)


Expected: the cursor can be placed at the center of the upper frame and at the center of the button in the lower frame.

Actual: the cursor can be placed at the center of the upper frame (good) and is placed near the upper left of the lower frame (bad).

I interpret this behavior as the cursor in the bottom frame being positioned not where it is expected but instead at the position the button would be in if that frame did not middle align its content.

**Comment by focustense** [comment]
Looking at the example, this has to do with two separate issues, which are both a little different from how it's presented.

1. Default focus when opening a menu. This is something that can be controlled at the layout level (i.e. in the code of a standard or custom view) but is difficult to present in an unambiguous, non-confusing way in StarML. However, the behavior if layouts don't customize is to find the first focusable element in the tree, which would be correct for something like this. The real _root cause_ here isn't with the default focus itself but with the method of testing, which sounds like using a mouse/keyboard to open the menu and then switching to (simulated) gamepad to try to navigate. Thus, the cursor does not snap to default focus because that is explicitly disabled when gamepad controls aren't in use. It's the _switch_ from mouse to gamepad controls that's preventing default focus from going to the right place, and that's not the most realistic of scenarios from a user's perspective; if they are a gamepad user, they'd open the menu in the first place using some gamepad mechanism.

2. Focus cannot reach inside the focusable element in a focus search because there is no straight-line path from the cursor; none of the cardinal directions lead to a point inside the inner frame. There is no straightforward solution to this problem no matter what kind of searching mechanism is used, even the "neighbor" system in vanilla UI.

Essentially, both of the problems _as stated_ are actually WAI, but there is a meta-problem to be solved here that doesn't have its own separate issue, so we can use this one. That problem is one of "lost focus", i.e. the scenario where (a) the cursor is not already in the bounds of some focusable UI element and also (b) a focus search in some direction, or possibly all directions, turns up empty.

In that scenario, there should be a way to "recover" focus, to get to the pointer to _someplace_ focusable (probably the default focusable element, as in point 1 above), even if it's not necessarily the "perfect" solution relative to where the mouse cursor is, just to get gamepad navigation back into a functional state; the assumption of course being that the author has designed the menu such that there aren't unreachable or inescapable "islands", so as long as the cursor ends up somewhere valid after a "last resort" search, the rest of the menu eventually becomes reachable.

It probably sounds a bit tacky, but in reality it's exactly what users of Stardew are going to be accustomed to because there isn't a real focus system in the first place, only cursor position and "snappy menus". So we are actually always trying to deduce focus from cursor position and sometimes the answer will be ambiguous or even counterintuitive - but it's generally "good enough" as long as we don't get outright _stuck_.

**Comment by pneuma163** [comment]
It sounds like you understand the underlying problem sufficiently but I thought I'd add one more piece of input in case it's helpful and in case there's still a separate issue present. You mentioned my method of testing being part of the cause of the issue - I had indeed opened the menu with a key press - so I changed the test code to open the menu using a DelayedAction triggered by a Warped event so that it was as disconnected from being a mouse/keyboard user as possible for my own setup. See the following 10s video for reference:

https://github.com/user-attachments/assets/3b7ca2b9-5611-46e9-98bf-a23b21924fa1

You also mentioned that "there is no straight-line path from the cursor" and I'm not certain whether you meant a straight line from the cursor's location when the menu opens (in that video it happens to be where I left it when I stopped touching my mouse a few seconds before the video) or a straight line between elements of the menu.

If it's the latter then the two *expected* focusable points would be the two centers of the two stacked frames (the top point given by the center of the frame, the bottom point given by the center of the button) and that seems like a different issue than focus recovery, particularly since those two points are precisely along the same vertical line, contrary to the observed behavior when switching between the two (i.e. we observe the cursor moving between the top's center and the bottom's upper left). If it's the former (i.e. potentially because of my method of testing without a controller) then I'll say I think I understand why you updated this issue to mention recovering focus.

For example, if I open the player inventory (in my "forced on" fake controller setup) then the focus snaps to the first spot in the inventory - it sounds like that's the issue you're recognizing as the problem here, that StardewUI menus do not do this focus recovery. However if there's an issue with navigation within a menu caused by the alignment properties, I wanted to include this additional comment to try to highlight it.

**Comment by focustense** [comment]
On my current machine, the emulated gamepad behaves very strangely even for vanilla menus, and the cursor snap position is always off from its expected position. Not sure if that's related in any way to the strange positioning here, and unfortunately, I can't test with a real gamepad right now - I actually have the gamepad, but found out when I tried to connect it that it doesn't work automatically and there's no current driver for it. Which was the only reason I attempted the emulated gamepad in the first place. So it's going to be a while longer before I can confirm anything related to focus.

I definitely agree that the center point should be where it snaps; I have no idea why it is not, except possibly for the emulated gamepad being screwy as mentioned above.
